src/class_loader_core.cpp

```diff
@@ -530,35 +523,28 @@ void unloadLibrary(const std::string & library_path, ClassLoader * loader)
     LibraryVector & open_libraries = getLoadedLibraryVector();
     LibraryVector::iterator itr = findLoadedLibrary(library_path);
     if (itr != open_libraries.end()) {
-      Poco::SharedLibrary * library = itr->second;
+      auto library = itr->second;
       std::string library_path = itr->first;
-      try {
-        destroyMetaObjectsForLibrary(library_path, loader);
-
-        // Remove from loaded library list as well if no more factories associated with said library
-        if (!areThereAnyExistingMetaObjectsForLibrary(library_path)) {
-          CONSOLE_BRIDGE_logDebug(
-            "class_loader.impl: "
-            "There are no more MetaObjects left for %s so unloading library and "
-            "removing from loaded library vector.\n",
-            library_path.c_str());
-          library->unload();
-          assert(library->isLoaded() == false);
-          delete (library);
-          itr = open_libraries.erase(itr);
-        } else {
-          CONSOLE_BRIDGE_logDebug(
-            "class_loader.impl: "
-            "MetaObjects still remain in memory meaning other ClassLoaders are still using library"
-            ", keeping library %s open.",
-            library_path.c_str());
-        }
-        return;
-      } catch (const Poco::RuntimeException & e) {
-        delete (library);
-        throw class_loader::LibraryUnloadException(
-                "Could not unload library (Poco exception = " + std::string(e.message()) + ")");
```

Comment:
The case that the library couldn't be unloaded (signaled by the return code of `rcutils_unload_shared_library`) isn't being handled anymore.

Reply:
I'm using here the destructor `library.reset();`. The destructor should not throw exceptions https://github.com/ros2/rcpputils/blob/master/src/shared_library.cpp#L44

Reply:
Imo that means the API is broken if it is unable to signal an error case properly.

Reply:
I can add a method to unload the library that will throws the exception.

Reply:
Added here https://github.com/ros2/rcpputils/pull/50